utils/histogram.hh

```diff
@@ -264,17 +265,101 @@ public:
         return res;
     }
 
-    void update() {
+    virtual void update() noexcept {
         for (int i = 0; i < 3; i++) {
             rates[i].update();
         }
     }
 
+    virtual ~timed_rate_moving_average() = default;
     uint64_t count() const {
         return _count;
     }
 };
 
+/*!
+ * \brief a class that holds a moving average rate, an estimated histogram, and a summary.
+ *
+ * A summary is a histogram where each bucket holds some quantile.
+ * While a histogram typically holds values from the system start,
+ * a summary is defined over some duration (i.e., latencies in the last 10 seconds).
+ * To calculate a summary, we use two estimated-histograms, calculate their delta, and get the
+ * summary from that delta histogram.
+ *
+ * timed_rate_moving_average_with_summary based on the timer mechanism in timed_rate_moving_average
+ * On each time tick, a summary is calculated and stored locally and the
+ * estimated histogram is updated.
+ */
+class timed_rate_moving_average_with_summary : public timed_rate_moving_average {
```

Comment:
I think composition is better than inheritance here. It avoids using virtual functions, and I'm guessing there will be no problem calling timed_rate_moving_average_with_summary.get_timed_rate_moving_average().

Reply:
Right.

I think it makes sense to move the timer out and leave this as a pure data+math class with no mechanics.

A (very small) benefit is that it allows consolidating timers, a single timer can update many moving averages. But it's good to have such classes focus on the math and not setup/teardown etc.